dbms/src/Storages/StorageKafka.cpp

```diff
@@ -0,0 +1,358 @@
+    if (consumer == nullptr)
+        throw Exception("Failed to create consumer handle: " + String(errstr), ErrorCodes::UNKNOWN_EXCEPTION);
+
+    rd_kafka_poll_set_consumer(consumer);
+
+    // Create a list of partitions
+    auto * topicList = rd_kafka_topic_partition_list_new(topics.size());
+    for (const auto & t : topics)
+    {
+        LOG_TRACE(log, "Subscribing to topic: " + t);
+        rd_kafka_topic_partition_list_add(topicList, t.c_str(), RD_KAFKA_PARTITION_UA);
+    }
+
+    // Subscribe to requested topics
+    auto err = rd_kafka_subscribe(consumer, topicList);
+    if (err)
+        throw Exception("Failed to subscribe: " + String(rd_kafka_err2str(err)), ErrorCodes::UNKNOWN_EXCEPTION);
+
+    rd_kafka_topic_partition_list_destroy(topicList);
+
+    // Start the reader thread
+    stream_thread = std::thread(&StorageKafka::streamThread, this);
+}
+
+
+void StorageKafka::shutdown()
+{
+    is_cancelled = true;
+    cancel_event.set();
+
+    LOG_TRACE(log, "Unsubscribing from assignments");
+    rd_kafka_unsubscribe(consumer);
+    auto err = rd_kafka_consumer_close(consumer);
+    if (err)
+    {
+        LOG_ERROR(log, "Failed to close: " + String(rd_kafka_err2str(err)));
+    }
+
+    LOG_TRACE(log, "Destroying consumer");
+    rd_kafka_destroy(consumer);
+    if (stream_thread.joinable())
+        stream_thread.join();
+
+    rd_kafka_wait_destroyed(CLEANUP_TIMEOUT_MS);
+}
+
+
+void StorageKafka::updateDependencies()
+{
+    cancel_event.set();
+}
+
+
+void StorageKafka::streamThread()
+{
+    setThreadName("KafkaStreamThread");
+
+    while (!is_cancelled)
+    {
+        try
+        {
+            auto dependencies = context.getDependencies(database_name, table_name);
+            if (dependencies.size() > 0)
+            {
+                LOG_DEBUG(log, "Started streaming to " << dependencies.size() << " attached views");
+                streamToViews();
+                LOG_DEBUG(log, "Stopped streaming to views");
+            }
+        }
+        catch (...)
+        {
+            tryLogCurrentException(__PRETTY_FUNCTION__);
+        }
+
+        cancel_event.tryWait(READ_POLL_MS);
+    }
+
+    LOG_DEBUG(log, "Stream thread finished");
+}
+
+
+void StorageKafka::streamToViews()
+{
+    auto table = context.getTable(database_name, table_name);
+    if (!table)
+    {
+        LOG_WARNING(log, "Destination table " << database_name << "." << table_name << " doesn't exist.");
+        return;
+    }
+
+    // Create an INSERT query for streaming data
+    auto insert = std::make_shared<ASTInsertQuery>();
+    insert->database = database_name;
+    insert->table = table_name;
+    insert->no_destination = true; // Only insert into dependent views
+
+    // Limit the number of batched messages to allow early cancellations
+    const Settings & settings = context.getSettingsRef();
+    const size_t block_size = std::min(settings.max_block_size.value, BATCH_SIZE_MAX);
```

Comment:
The size is capped as the underlying library applies timeout for each received message individually. With upper bound of 16, the maximum time spent blocked would be 16 seconds (worst case scenario is messages arriving in at stead rate of slightly less than 1 second).